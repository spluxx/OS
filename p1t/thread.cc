#include <stdlib.h>
#include <stdio.h>
#include <iostream>
#include <ucontext.h>
#include <queue>
#include <map>
#include <algorithm>
#include "thread.h"
#include "interrupt.h"
using namespace std;

#ifdef DEBUG
  #define DEBUG(M, ...) fprintf(stderr, "[DEBUG] %s:%d: " M "\n", __FILE__, __LINE__, ##__VA_ARGS__)
#else
  #define DEBUG(M, ...)
#endif

typedef unsigned int lock_t;
typedef pair<unsigned int, unsigned int> lock_cv_t;

static int complete_threads;
static map<ucontext_t *, bool> thread_complete; // true if complete

// lock owners
// if available, owner = null
static map<lock_t, ucontext_t *> lock_owner;

// ucontext storages
static ucontext_t *running, *tmp; 
static deque<ucontext_t *> readyQ;
static map<lock_t, deque<ucontext_t *> > wait_lockQ;
static map<lock_cv_t, deque<ucontext_t *> > wait_cvQ;

//------------------------helper functions-------------------------------//

int interrupt_enable(int ret) { interrupt_enable(); return ret; }
int interrupt_disable(int ret) { interrupt_disable(); return ret; }

void free_ucontext(void *ucp) {
  free(((ucontext_t *)ucp)->uc_stack.ss_sp);
  free(ucp);
}

void collect_garbage(bool done) {
  map<ucontext_t *, bool>::iterator it = thread_complete.begin();
  if(done) free(tmp);
  while(it != thread_complete.end()) {
    // if the program's about to exit OR if the thread is completed
    if((it->first != running) && (done || it->second)) {
      free_ucontext(it->first); // free
      thread_complete.erase(it++); // erase from thread_complete
    } else it ++;
  }
  complete_threads = 0;
}

// function wrapper, forcing the given function to
// 1. run the function with given arguments
// 2. report completion for GC;
// 3. yield CPU to next ready thread.
void func_extend(void *ucp, thread_startfunc_t func, void *arg) {
  interrupt_enable();
  func(arg);
  thread_complete[(ucontext_t *)ucp] = true;
  complete_threads ++;
  thread_yield(); // after freeing whatever, yield the CPU
}

int exit_lib() {
  collect_garbage(true);
  cout << "Thread library exiting.\n";
  exit(0);
  return 0;
}

//------------------------library functions-------------------------------//

int thread_libinit(thread_startfunc_t func, void *arg) { 
  if(tmp != NULL) return -1;
  running = tmp = NULL;
  complete_threads = 0;
  tmp = (ucontext_t *) malloc(sizeof(ucontext_t));
  if(tmp == NULL) return -1;
  if(thread_create(func, arg) == -1) return -1; // create initial thread
  if(thread_yield() == -1) return -1; // start the initial thread
  return -1; // NOT EXECUTED
}

int thread_create(thread_startfunc_t func, void *arg) {
  interrupt_disable();
  if(tmp == NULL) return interrupt_enable(-1);
  void *stk_ptr = malloc(STACK_SIZE); 
  ucontext_t *ucp = (ucontext_t *) malloc(sizeof(ucontext_t));

  if(stk_ptr == NULL || ucp == NULL) {
    free(stk_ptr);
    free(ucp);
    return interrupt_enable(-1);
  }

  // setup context for the new thread
  if(getcontext(ucp) == -1) return -1;
  ucp->uc_stack.ss_sp = stk_ptr;
  ucp->uc_stack.ss_size = STACK_SIZE;
  ucp->uc_stack.ss_flags = 0;
  ucp->uc_link = NULL; 

  makecontext(ucp, (void (*)()) func_extend, 3, ucp, func, arg); 

  thread_complete[ucp] = false;
  readyQ.push_back(ucp);
  return interrupt_enable(0);
}

int thread_yield(void) { 
  interrupt_disable();
  if(tmp == NULL) return interrupt_enable(-1);
  
  if(readyQ.size() == 0) {
    if(thread_complete[running]) exit_lib();
    else return interrupt_enable(0);
  } 

  if(complete_threads > 10) collect_garbage(false);
  
  bool init_thread = running == NULL;
  // DO NOT put back into ready queue 
  // if the current running thread is
  // 1. the initial thread (running = NULL)
  // 2. done running
  bool noswap = init_thread || thread_complete[running];
  if(!noswap) readyQ.push_back(running); 

  // pop a thread to run, setting it as running
  running = readyQ.front(); 
  readyQ.pop_front(); 
  if(swapcontext(noswap ? tmp : readyQ.back(), running) == -1) return -1;
  if(!init_thread) {
    interrupt_enable();
  } return 0;
}

int thread_lock(lock_t lock) {
  interrupt_disable();
  if(tmp == NULL) return interrupt_enable(-1);

  // if lock doesn't even exist within our knowledge, initialize it to NULL
  if(lock_owner.find(lock) == lock_owner.end()) lock_owner[lock] = NULL;
  // if the thread wants a lock when it already holds it - fuck you
  if(lock_owner[lock] == running) return interrupt_enable(-1);
  // if the lock is unavailable, wait
  if(lock_owner[lock] != NULL) {
    wait_lockQ[lock].push_back(running);
    if(readyQ.size() == 0) { interrupt_enable(); exit_lib(); }
    running = readyQ.front();
    readyQ.pop_front();
    if(swapcontext(wait_lockQ[lock].back(), running) == -1) return -1;
  } else lock_owner[lock] = running;
  return interrupt_enable(0);
}

int thread_unlock(lock_t lock) {
  interrupt_disable();
  if(tmp == NULL) return interrupt_enable(-1);
  // if lock doesn't even exist within our knowledge, initialize it to NULL
  if(lock_owner.find(lock) == lock_owner.end()) lock_owner[lock] = NULL;
  // if the thread tries to return a lock it doesn't even own - fuck you
  if(lock_owner[lock] != running) return interrupt_enable(-1);
  // alright this thread actually has a lock - unlock it;
  if(wait_lockQ[lock].size() > 0) {
    lock_owner[lock] = wait_lockQ[lock].front();
    readyQ.push_back(wait_lockQ[lock].front());
    wait_lockQ[lock].pop_front();
  } else lock_owner[lock] = NULL;
  return interrupt_enable(0);
}

int thread_wait(lock_t lock, lock_t cv) {
  interrupt_disable();
  if(tmp == NULL) return interrupt_enable(-1);

  lock_cv_t lock_cv = make_pair(lock, cv);
  // if lock doesn't even exist within our knowledge, initialize it to NULL
  if(lock_owner.find(lock) == lock_owner.end()) lock_owner[lock] = NULL;
  // if the thread doesn't have a lock to begin with - fuck you
  if(lock_owner[lock] != running) return interrupt_enable(-1);
  
  // drop the lock
  if(wait_lockQ[lock].size() > 0) {
    lock_owner[lock] = wait_lockQ[lock].front();
    readyQ.push_back(wait_lockQ[lock].front());
    wait_lockQ[lock].pop_front();
  } else lock_owner[lock] = NULL;

  // and wait
  wait_cvQ[lock_cv].push_back(running);

  // yield
  if(readyQ.size() == 0) { interrupt_enable(); exit_lib(); } 
  running = readyQ.front();
  readyQ.pop_front();
  if(swapcontext(wait_cvQ[lock_cv].back(), running) == -1) return -1;
  interrupt_enable();

  // when it wakes up, pick up the lock
  return thread_lock(lock);
}

int thread_signal(lock_t lock, lock_t cv) {
  interrupt_disable();
  if(tmp == NULL) return interrupt_enable(-1);
  lock_cv_t lock_cv = make_pair(lock, cv);

  // if lock doesn't even exist within our knowledge, initialize it to NULL
  if(lock_owner.find(lock) == lock_owner.end()) lock_owner[lock] = NULL;
  
  // wake a waiting thread up
  if(wait_cvQ[lock_cv].size() == 0) return interrupt_enable(0);
  readyQ.push_back(wait_cvQ[lock_cv].front());
  wait_cvQ[lock_cv].pop_front();
  
  return interrupt_enable(0);
}

int thread_broadcast(lock_t lock, lock_t cv) {
  interrupt_disable();  
  if(tmp == NULL) return interrupt_enable(-1);
  lock_cv_t lock_cv = make_pair(lock, cv);

  // if lock doesn't even exist within our knowledge, initialize it to NULL
  if(lock_owner.find(lock) == lock_owner.end()) lock_owner[lock] = NULL;
  
  // wake every waiting thread up
  int size = wait_cvQ[lock_cv].size();
  for(int i = 0 ; i < size; i ++) {
    readyQ.push_back(wait_cvQ[lock_cv].front());
    wait_cvQ[lock_cv].pop_front();
  }
  
  return interrupt_enable(0);
}
