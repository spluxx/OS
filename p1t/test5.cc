#include <stdio.h>
#include <stdlib.h>
#include <iostream>
#include <vector>
#include "thread.h"
using namespace std;

int abs(int a) { return a > 0 ? a : -a; }

// ------- Locks -------------------------
const int BOARD_LOCK = 1; //		  |
const int NUMBER_OF_CASHIER_LOCK = 2; //  |
const int MONITOR_LOCK = 3;
// ------- Condition Variables -----------
const int BOARD_FULL = 4; //		  |
const int BOARD_NOT_FULL = 5; //	  |
// ---------------------------------------


struct order_info {
  int cashier_id;
  int sandwich_id;
  order_info(){}
  order_info(int c, int s) { 
    cashier_id = c;
    sandwich_id = s;
  }
};

int max_orders, nCashier; 
bool full, done;

// GOVERNED BY NUMBER_OF_CASHIER_LOCK --------------
int liveCashier;			    //      |
// -------------------------------------------------

// GOVERNED BY BOARD_LOCK ---------------------------
vector<order_info> board; //			     | 
int *order_complete; // 0 (incomplete) 1 (complete)  |
// --------------------------------------------------

void create_maker() {
  int made_last = -1;

  while(true) {
    int mi = 0;
    int mn = 5000;
    order_info toCook;

    thread_lock(BOARD_LOCK);
    // wait for board_full signal
    while(!full) thread_wait(BOARD_LOCK, BOARD_FULL);	
    if(done) break;

    // find closest sandwich from made_last
    for(int i = 0 ; i < board.size() ; i ++) {		
      int diff = abs(board[i].sandwich_id - made_last);
      if(diff < mn) {
        mi = i;
        mn = diff;
        toCook = board[i];
      }
    }

    board.erase(board.begin()+mi); // remove from board
    made_last = toCook.sandwich_id; // update latest sandwich made
    order_complete[toCook.cashier_id] = 1; // cooking completed!

    thread_lock(MONITOR_LOCK);
    cout << "READY: cashier " << toCook.cashier_id << " sandwich " << toCook.sandwich_id << endl; 
    thread_unlock(MONITOR_LOCK);

    full = false;
    thread_broadcast(BOARD_LOCK, BOARD_NOT_FULL); // broadcast board_not_full signal

    thread_unlock(BOARD_LOCK);
  }

  free(order_complete);
}

void create_cashier(int* sandwiches) {
  int myID = -1;

  thread_lock(NUMBER_OF_CASHIER_LOCK); 
  myID = liveCashier++;		       
  thread_unlock(NUMBER_OF_CASHIER_LOCK);
  
  int sandwich = -1;
  order_complete[myID] = 1;

  // read order one by one
  for(int i = 0 ; i < 5 ; i ++) {
    sandwich = sandwiches[i];
    thread_lock(BOARD_LOCK); 							

    // the previous order must have been completed
    // to wait for board_not_full broadcast 
    while(full || order_complete[myID] == 0) thread_wait(BOARD_LOCK, BOARD_NOT_FULL);

    board.push_back(order_info(myID, sandwich));	  // post an order
    order_complete[myID] = 0;

    thread_lock(MONITOR_LOCK);
    cout << "POSTED: cashier " << myID << " sandwich " << sandwich << endl; 
    thread_unlock(MONITOR_LOCK);

    if(board.size() == max_orders) {
      full = true; 
      // send board_full signal
      thread_signal(BOARD_LOCK, BOARD_FULL);
    }

    thread_unlock(BOARD_LOCK);
  }

  // wait for final order
  thread_lock(BOARD_LOCK);
  while(order_complete[myID] == 0) thread_wait(BOARD_LOCK, BOARD_NOT_FULL);
  thread_unlock(BOARD_LOCK);

  // cashier can go home now
  thread_lock(NUMBER_OF_CASHIER_LOCK); 
  if(--liveCashier < max_orders) {
    max_orders = liveCashier;  
    if(liveCashier == 0) done = true;

    thread_lock(BOARD_LOCK);
    full = true;
    thread_signal(BOARD_LOCK, BOARD_FULL);
    thread_unlock(BOARD_LOCK);
  }
  thread_unlock(NUMBER_OF_CASHIER_LOCK);
}

void initialize(void) {
  order_complete = (int *) malloc(sizeof(int)*nCashier);
  int sandwiches[5][5] ={
    {1, 3, 53, 100, 50},
    {5, 530, 123, 539, 929},
    {19, 512, 12, 942, 42},
    {63, 623, 236, 123, 643},
    {464, 235, 132, 696, 709}
  };
  int** sandwich_number = (int **) malloc(sizeof(int *) * 5);
  for(int i = 0 ; i < 5 ; i ++) {
    sandwich_number[i] = (int *) malloc(sizeof(int) * 5);
    for(int j = 0 ; j < 5 ; j ++)
      sandwich_number[i][j] = sandwiches[i][j];
  }

  for(int i = 0 ; i < 5 ; i ++)
    thread_create((thread_startfunc_t) create_cashier, (void*) sandwich_number[i]);
  thread_create((thread_startfunc_t) create_maker, NULL);
}

int main(int argc, char* argv[]) {
  nCashier = 5;
  max_orders = 3;
  liveCashier = 0;
  full = false;
  done = nCashier <= 0;
  max_orders = max_orders < nCashier ? max_orders : nCashier;

  thread_libinit((thread_startfunc_t) initialize, NULL);
  return 0;
}
