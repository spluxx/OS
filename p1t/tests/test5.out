POSTED: cashier 0 sandwich 1
POSTED: cashier 1 sandwich 5
POSTED: cashier 2 sandwich 19
READY: cashier 0 sandwich 1
POSTED: cashier 0 sandwich 3
READY: cashier 0 sandwich 3
POSTED: cashier 0 sandwich 53
READY: cashier 1 sandwich 5
POSTED: cashier 1 sandwich 530
READY: cashier 2 sandwich 19
POSTED: cashier 2 sandwich 512
READY: cashier 0 sandwich 53
POSTED: cashier 0 sandwich 100
READY: cashier 0 sandwich 100
POSTED: cashier 0 sandwich 50
READY: cashier 0 sandwich 50
POSTED: cashier 3 sandwich 63
READY: cashier 3 sandwich 63
POSTED: cashier 3 sandwich 623
READY: cashier 2 sandwich 512
POSTED: cashier 2 sandwich 12
READY: cashier 1 sandwich 530
POSTED: cashier 1 sandwich 123
READY: cashier 3 sandwich 623
POSTED: cashier 3 sandwich 236
READY: cashier 3 sandwich 236
POSTED: cashier 3 sandwich 123
READY: cashier 1 sandwich 123
POSTED: cashier 1 sandwich 539
READY: cashier 3 sandwich 123
POSTED: cashier 3 sandwich 643
READY: cashier 2 sandwich 12
POSTED: cashier 2 sandwich 942
READY: cashier 1 sandwich 539
POSTED: cashier 1 sandwich 929
READY: cashier 3 sandwich 643
POSTED: cashier 4 sandwich 464
READY: cashier 4 sandwich 464
POSTED: cashier 4 sandwich 235
READY: cashier 4 sandwich 235
POSTED: cashier 4 sandwich 132
READY: cashier 4 sandwich 132
POSTED: cashier 4 sandwich 696
READY: cashier 4 sandwich 696
POSTED: cashier 4 sandwich 709
READY: cashier 4 sandwich 709
READY: cashier 1 sandwich 929
READY: cashier 2 sandwich 942
POSTED: cashier 2 sandwich 42
READY: cashier 2 sandwich 42
Thread library exiting.
